lib/concurrent/atomic/read_write_lock.rb

```diff
@@ -0,0 +1,168 @@
+require 'thread'
+require 'concurrent/atomic'
+
+module Concurrent
+
+  # Ruby read-write lock implementation
+  #
+  # Allows any number of concurrent readers, but only one concurrent writer
+  # (And if the "write" lock is taken, any readers who come along will have to wait)
+  #
+  # If readers are already active when a writer comes along, the writer will wait for
+  # all the readers to finish before going ahead.
+  # Any additional readers that come when the writer is already waiting, will also
+  # wait (so writers are not starved).
+  #
+  # @example
+  #   lock = Concurrent::ReadWriteLock.new
+  #   lock.with_read_lock  { data.retrieve }
+  #   lock.with_write_lock { data.modify! }
+  #
+  # @note Do **not** try to acquire the write lock while already holding a read lock
+  #   **or** try to acquire the write lock while you already have it.
+  #   This will lead to deadlock
+  class ReadWriteLock
+
+    WAITING_WRITER  = 1 << 15
+    RUNNING_WRITER  = 1 << 30
+    MAX_READERS     = WAITING_WRITER - 1
+    MAX_WRITERS     = RUNNING_WRITER - MAX_READERS - 1
+
+    # Implementation notes: 
+    # A goal is to make the uncontended path for both readers/writers lock-free
+    # Only if there is reader-writer or writer-writer contention, should locks be used
+    # Internal state is represented by a single integer ("counter"), and updated 
+    #  using atomic compare-and-swap operations
+    # When the counter is 0, the lock is free
+    # Each reader increments the counter by 1 when acquiring a read lock
+    #   (and decrements by 1 when releasing the read lock)
+    # The counter is increased by (1 << 15) for each writer waiting to acquire the
+    #   write lock, and by (1 << 30) if the write lock is taken
+
+    def initialize
+      @counter      = Atomic.new(0)         # single integer which represents lock state
+      @reader_q     = ConditionVariable.new # queue for waiting readers
+      @reader_mutex = Mutex.new             # to protect reader queue
+      @writer_q     = ConditionVariable.new # queue for waiting writers
+      @writer_mutex = Mutex.new             # to protect writer queue
+    end
+
+    def with_read_lock
+      acquire_read_lock
+      result = yield
+      release_read_lock
+      result
+    end
+
+    def with_write_lock
+      acquire_write_lock
+      result = yield
+      release_write_lock
+      result
+    end
+
+    def acquire_read_lock
+      while(true)
+        c = @counter.value
+        raise "Too many reader threads!" if (c & MAX_READERS) == MAX_READERS
+
+        # If a writer is waiting when we first queue up, we need to wait
+        if c >= WAITING_WRITER
+          # But it is possible that the writer could finish and decrement @counter right here...
+          @reader_mutex.synchronize do 
+            # So check again inside the synchronized section
+            @reader_q.wait(@reader_mutex) if @counter.value >= WAITING_WRITER

```

Comment:
It doesn't matter if `@counter` is updated elsewhere when a reader is in this `synchronized` block. `@counter` is always updated with an atomic CAS, so updates to `@counter` can never be lost.

The point of this line is so that a reader doesn't sleep forever if it tries to acquire the lock at the moment a writer is releasing it. (Remember, the only thing that can make a reader sleep is if a writer already holds the lock. Readers don't block readers.)

Now, when a writer releases the lock, it acquires `@reader_mutex` before waking up the waiting readers. If the lock-acquiring reader gets into this `synchronized` block before the lock-releasing writer, then the reader will go to sleep and atomically release `@reader_mutex`, then the leaving writer will get into its `synchronized` block and wake the reader up. On the other hand, if the leaving writer gets into its `synchronized` block first, `@counter` will already be updated and visible to all threads by that time the reader gets into its `synchronized` block, the condition will be false and it will not sleep. In either case, the reader will not sleep forever.